### PR TITLE
Tests for binascii

### PR DESCRIPTION
**Pull request by akash-suresh**
I have added tests for the `binascii` module. Let me know if there are any corrections needed.

### CODE REVIEW

File: tests/test_encode_decode.py [hunk 1]
```diff
@@ -95,8 +96,80 @@ def test_b85_encode_decode_round_trip(self, payload, pad):
...
+
+    @given(
+        payload_piece_1=st.binary(),
+        payload_piece_2=st.binary(),
+        value=st.just(0) | st.integers(),
+    )
+    def test_crc_hqx_two_pieces(self, payload_piece_1, payload_piece_2, value):
+        combined_crc = binascii.crc_hqx(payload_piece_1 + payload_piece_2, value)
+        crc = binascii.crc_hqx(payload_piece_1, value)
+        crc = binascii.crc_hqx(payload_piece_2, crc)
```
Comment (Zac-HD) [inline tests/test_encode_decode.py]:
```suggestion
        crc_part1 = binascii.crc_hqx(payload_piece_1, value)
        crc = binascii.crc_hqx(payload_piece_2, crc_part1)
```

make it clear that we're not discarding the variable we just defined.

File: tests/test_encode_decode.py [hunk 2]
```diff
@@ -95,8 +96,80 @@ def test_b85_encode_decode_round_trip(self, payload, pad):
...
+
+    @given(
+        payload_piece_1=st.binary(),
+        payload_piece_2=st.binary(),
+        value=st.just(0) | st.integers(),
+    )
+    def test_crc32_two_part(self, payload_piece_1, payload_piece_2, value):
+        combined_crc = binascii.crc32(payload_piece_1 + payload_piece_2, value)
+        crc = binascii.crc32(payload_piece_1, value)
+        crc = binascii.crc32(payload_piece_2, crc)
```
Comment (Zac-HD) [inline tests/test_encode_decode.py]:
```suggestion
        crc_part1 = binascii.crc32(payload_piece_1, value)
        crc = binascii.crc32(payload_piece_2, crc_part1)
```

File: tests/test_encode_decode.py [hunk 3]
```diff
@@ -95,8 +96,80 @@ def test_b85_encode_decode_round_trip(self, payload, pad):
...
+
+    @given(payload=st.binary())
+    def test_rlecode_hqx_rledecode_hqx_round_trip(self, payload):
+        x = binascii.rlecode_hqx(payload)
+        self.assertEqual(payload, binascii.rledecode_hqx(x))
+
+    @given(payload=st.binary())
+    def test_b2a_hqx_a2b_hqx_round_trip(self, payload):
+        # assuming len(payload) as 3, since it throws exception: binascii.Incomplete, when length is not a multiple of 3
+        assume(len(payload) % 3 == 0)
```
Comment (Zac-HD) [inline tests/test_encode_decode.py]: Let's actually test that:

```suggestion
        if len(payload) % 3:
            with self.assertRaises(binascii.Incomplete):
                binascii.b2a_hqx(payload)
            payload += b"\x00" * (len(payload) % 3)
```

File: tests/test_encode_decode.py [hunk 4]
```diff
@@ -95,8 +96,80 @@ def test_b85_encode_decode_round_trip(self, payload, pad):
 
 
 class TestBinASCII(unittest.TestCase):
-    # TODO: https://docs.python.org/3/library/binascii.html
-    pass
+    @given(payload=st.binary(), backtick=st.booleans())
+    def test_b2a_uu_a2b_uu_round_trip(self, payload, backtick):
+        x = binascii.b2a_uu(payload, backtick=backtick)
```
Comment (Zac-HD) [inline tests/test_encode_decode.py]: [The `backtick` option is new in Python 3.7](https://docs.python.org/3/library/binascii.html#binascii.b2a_uu), so we'll want a condition like `if sys.version_info[:2] >= (3, 7):` and only pass it in the first branch.

**Review by Zac-HD — COMMENTED** [review]
Nice work @akash-suresh!

**Comment by Zac-HD** [comment]
Cross-reference: https://github.com/HypothesisWorks/hypothesis/issues/2430

File: tests/test_encode_decode.py (hunk 4 above)
Comment (akash-suresh) [reply tests/test_encode_decode.py]: Sure 👍

**Comment by akash-suresh** [comment]
Hey @Zac-HD! I have made changes, based on the comments.

**Comment by Zac-HD** [comment]
```python
import binascii
binascii.crc_hqx(b"", 2 ** 63)
```

fails (only) on PyPy, with `OverflowError: int too large to convert to int`.  I think you've found a bug!

@cfbolz might be able to help us diagnose this?

File: tests/test_encode_decode.py [hunk 5]
```diff
@@ -95,8 +97,92 @@ def test_b85_encode_decode_round_trip(self, payload, pad):
...
+    def test_b2a_uu_a2b_uu_round_trip(self, payload):
+        x = binascii.b2a_uu(payload)
+        self.assertEqual(payload, binascii.a2b_uu(x))
+
+    @unittest.skipIf(
+        sys.version_info[:2] < (3, 7), "backtick not supported in this library version"
+    )
+    @given(payload=st.binary(), backtick=st.booleans())
+    def test_b2a_uu_a2b_uu_round_trip_with_backtick(self, payload, backtick):
+        x = binascii.b2a_uu(payload, backtick=backtick)
```
Comment (Zac-HD) [inline tests/test_encode_decode.py]:
```suggestion
    @given(payload=st.binary(), backtick=st.booleans())
    def test_b2a_uu_a2b_uu_round_trip_with_backtick(self, payload, backtick):
        if sys.version_info[:2] >= (3, 7):
            x = binascii.b2a_uu(payload, backtick=backtick)
        else:
            x = binascii.b2a_uu(payload)
```

we can still run the test on PyPy's 3.6 implementation, we just ignore the `backtick` argument.

Comment (akash-suresh) [reply tests/test_encode_decode.py]: Sure 👍

File: tests/test_encode_decode.py [hunk 6]
```diff
@@ -95,8 +97,87 @@ def test_b85_encode_decode_round_trip(self, payload, pad):
...
+        if len(payload) % 3:
+            with self.assertRaises(binascii.Incomplete):
+                x = binascii.b2a_hqx(payload)
+                binascii.a2b_hqx(x)
+            payload += b"\x00" * (-len(payload) % 3)
+        x = binascii.b2a_hqx(payload)
+        res, _ = binascii.a2b_hqx(x)
+        self.assertEqual(payload, res)
+
+    @given(payload=st.binary(), value=st.just(0) | st.integers())
```
Comment (Zac-HD) [inline tests/test_encode_decode.py:140]: Final suggestion - let's preserve the pypy issue you found by adding `@example(payload=b"", value=2 ** 63)`.  You'll need to rebase on master for this, `example` is already imported on my branch.

After that I'm happy to merge once PyPy developers have seen this.

Comment (akash-suresh) [reply tests/test_encode_decode.py:140]: Sure 👍 

I had a doubt, just to be sure. What exactly do you mean by, rebase on master?

Comment (Zac-HD) [reply tests/test_encode_decode.py:140]: https://www.atlassian.com/git/tutorials/rewriting-history and https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase are good introductions :slightly_smiling_face: 

This is intermediate OSS contributor level - I would normally just fix up the history after a new contributor is done, but you've been going really well and so I thought you might be up for learning another new skill!

**Comment by cfbolz** [comment]
@Zac-HD nice catch! yes, will look into it!

**Comment by cfbolz** [comment]
yep, fixed (an unsigned vs signed int issue, also a change from py2->py3). here's the commit: https://foss.heptapod.net/pypy/pypy/commit/7b826e1a27677daa643bfa4535daafecc4526335
Are we keeping a list of found bugs yet? We should add this!

**Comment by Zac-HD** [comment]
@cfbolz, impressively fast!

We do have https://github.com/Zac-HD/stdlib-property-tests#trophy-case - is there a canonical link to point to for this, or should the new entry just point to this thread and commit link?

File: tests/test_encode_decode.py (hunk 6 above)
Comment (akash-suresh) [reply tests/test_encode_decode.py:140]: Hey @Zac-HD. Sorry to bother you with something this silly. But, I am getting a lot of conflicts while doing the rebase. Is that normal?

Comment (akash-suresh) [reply tests/test_encode_decode.py:140]: I figured it out. 😄

Comment (Zac-HD) [reply tests/test_encode_decode.py:140]: Not silly!  It really depends on what you're rebasing - sometimes I have no commits, sometimes so many that I just give up and try something else :sweat_smile: 

Sounds like you worked it out though, I can see you pushed the new version - nice work!

**Comment by cfbolz** [comment]
@Zac-HD I'd just point to the commit link - we don't track all changes as issues or anything like that.

**Pull request merged**